architecture/2024-08-controller-component-threat-model.md

```diff
@@ -0,0 +1,185 @@
+
+## Data Flow
+
+### Diagram
+
+![Controller Component via Microsoft Threat Modeling Tool](./2024-08-controller-component-threat-model/controller-component.png)
+
+1. **User creates/updates/deletes a Recipe or a Deployment resource**: When a user requests to create, update, or delete a Recipe or a Deployment resource, the request is handled by the Kubernetes API Server. One way, and probably the most common way, a user can do this request is by running a **kubectl** command. Kubernetes takes care of the authentication and the authorization of the user and its request(s) so we (Radius) don't need to worry about anything here.
+
+2. **Validating Webhook**: The only type of admission controller we have in Radius is the validating webhook for the Recipe resource. The validating webhook ensures that the Recipe object to be created or updated is one of the Radius portable resources. Whenever Kubernetes API Server receives a request to create or update a Recipe object, it communicates the proposed changes with the validating webhook. If the validating webhook validates the changes, then it is persisted to the **etcd** by the Kubernetes API Server.
+
+3. **Recipe and Deployment Reconcilers**: When there is a request to create, update, or delete a Recipe or a Deployment resource, after being validated if the resource is a Recipe resource, the next step is the reconcilation of the resource by the appropriate reconciler. In the Controller component, there are two reconcilers: Recipe and Deployment. These reconcilers are loops that watch the changes in the Recipe and Deployment resources. Whenever there is a change, the reconcilers take the necessary actions to move the actual state to the desired state. These necessary actions include communication with the UCPD to create, update, and/or delete necessary resources.
+
+   1. Communication:
+      1. Controller and UCPD:
+         1. Poll long-running operations (create, update, or delete) for a Recipe or a Deployment resource.
+         2. Create a Radius Resource Group if needed.
+         3. Create a Radius Application if needed.
+         4. Get a Radius resource like the Environment that is associated with the resource.
+         5. Create/Update/Delete a Recipe or a Deployment resource.
+         6. Create/Update/Delete a Secret for a Recipe or a Deployment resource.
+      2. Controller and the Kubernetes API Server:
+         1. Fetch a Recipe or a Deployment object.
+         2. Send events related to the operations running.
+         3. Update a Recipe or a Deployment object.
+         4. Create/Update/Delete a Secret associated with a Recipe or a Deployment object.
+         5. List Deployments by filtering them by a specific Recipe object.
+
+### Threats
+
+#### Threat: Users with access to the webhook server modifying the behavior of the webhook server
```

Comment:
I think this is still unclear. As we discussed in the design meeting, we're assuming that the user has not tampered with Radius.

> If the user wants to crash their own car, that's allowed 😆

Reply:
Let's sync up on this offline.